server/channels/api4/file.go

```diff
@@ -30,12 +33,87 @@ const (
 
 const maxMultipartFormDataBytes = 10 * 1024 // 10Kb
 
+// sendFileDownloadRejectedEvent sends a websocket event to notify the user that their file download was rejected
+func sendFileDownloadRejectedEvent(app *app.App, info *model.FileInfo, userID string, rejectionReason string, downloadType model.FileDownloadType) {
+	if userID == "" {
+		// Don't send event for public files (no user context)
+		app.Log().Debug("Skipping websocket event for public file download rejection")
+		return
+	}
+
+	message := model.NewWebSocketEvent(model.WebsocketEventFileDownloadRejected, "", info.ChannelId, userID, nil, "")
+	message.Add("file_id", info.Id)
+	message.Add("file_name", info.Name)
+	message.Add("rejection_reason", rejectionReason)
+	message.Add("channel_id", info.ChannelId)
+	message.Add("post_id", info.PostId)
+	message.Add("download_type", string(downloadType))
+	app.Publish(message)
+}
+
+// runFileWillBeDownloadedHook executes the FileWillBeDownloaded hook with a timeout
+// Returns empty string to allow download, or a rejection reason to block it.
+// If the hook times out, the download is blocked for security reasons.
+// downloadType should be one of model.FileDownloadType constants
+func runFileWillBeDownloadedHook(app *app.App, pluginContext *plugin.Context, fileInfo *model.FileInfo, userID string, downloadType model.FileDownloadType, translateFunc func(string, ...any) string) string {
+	// Check if plugins are enabled and environment is available
+	if app.GetPluginsEnvironment() == nil {
+		app.Log().Debug("Plugin environment not available, skipping FileWillBeDownloaded hook")
+		return ""
+	}
+
+	// Use configured timeout for file download hooks to prevent blocking user downloads
+	// and ensure security by blocking downloads if plugins are unresponsive
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(model.PluginSettingsDefaultHookTimeoutSeconds)*time.Second)
+	defer cancel()
+
+	var rejectionReason atomic.Value
+	done := make(chan struct{})
+
+	app.Srv().Go(func() {
+		defer close(done)
+		app.Channels().RunMultiHook(func(hooks plugin.Hooks, _ *model.Manifest) bool {
+			rejectionReasonFromHook := hooks.FileWillBeDownloaded(pluginContext, fileInfo, userID, downloadType)
+			rejectionReason.Store(rejectionReasonFromHook)
+			app.Log().Debug("FileWillBeDownloaded hook called",
```

Comment:
Left over from development?

Reply:
The debug string can be improved, but I left this debug line on purpose to have some data in the logs when we hit the hook.